lib/logstash/outputs/base.rb

```diff
@@ -54,32 +59,57 @@ def receive(event)
 
   public
   def worker_setup
+    # TODO: Remove this branch, delete this function
     if @workers == 1
       @worker_plugins = [self]
     else
-      define_singleton_method(:handle, method(:handle_worker))
-      @worker_queue = SizedQueue.new(20)
+      define_singleton_method(:handle_multi, method(:handle_worker))
+
+      @available_workers = SizedQueue.new(@worker_plugins.length)
+
       @worker_plugins = @workers.times.map { self.class.new(@original_params.merge("workers" => 1)) }
-      @worker_plugins.map.with_index do |plugin, i|
-        Thread.new(original_params, @worker_queue) do |params, queue|
-          LogStash::Util::set_thread_name(">#{self.class.config_name}.#{i}")
-          plugin.register
-          while true
-            event = queue.pop
-            plugin.handle(event)
-          end
-        end
+
+      @worker_plugins.each do |wp|
+        wp.register
+        @available_workers << wp
       end
     end
   end
 
   public
+  # Not to be overriden by plugin authors!
   def handle(event)
-    receive(event)
+    @single_worker_mutex.synchronize { receive(event) }
   end # def handle
 
-  def handle_worker(event)
-    @worker_queue.push(event)
+  # To be overriden in implementations
+  def receive_multi(events)
+    events.each {|event|
+      receive(event)
+    }
+  end
+
+  # Not to be overriden by plugin authors!
+  def handle_multi(events)
+    @single_worker_mutex.synchronize { receive_multi(events) }
+  end
+
+  def handle_worker(events)
```

Comment:
for performance sake (and maybe readability) I'd consider implementing this rotation with a simple array and a number pointer modulo @worker_plugins.size that grows per event or even using Array#cycle such as:

```ruby
@available_workers.cycle do |worker|
   worker.handle_multi(events)
end
```